segmentation_models_pytorch/encoders/timm_universal.py

```diff
@@ -23,34 +75,134 @@ def __init__(
             out_indices=tuple(range(depth)),
         )
 
-        # not all models support output stride argument, drop it by default
+        # Ｎot all models support output stride argument, drop it by default
         if output_stride == 32:
             common_kwargs.pop("output_stride")
 
-        self.model = timm.create_model(
-            name, **_merge_kwargs_no_duplicates(common_kwargs, kwargs)
-        )
+        # Load a temporary model to analyze its feature hierarchy
+        self.model = timm.create_model(name, features_only=True)
```

Comment:
Why do we need to load a temporary model? I would try to avoid it if possible.

Reply:
I believe that a temporary model is necessary because we need to determine `feature_info.reduction()` to classify the model as traditional, transformer, or VGG style. This affects the range of `out_indices` to be used:
```python
common_kwargs["out_indices"] = tuple(range(depth))
```
- If `depth == 5`, `out_indices` is
  - traditional-style (0, 1, 2, 3, 4)
  - transformer-style (0, 1, 2, 3)
  - vgg-style (0, 1, 2, 3, 4, 5)
- If `depth == 3`, `out_indices` is
  - traditional-style (0, 1, 2)
  - transformer-style (0, 1)
  - vgg-style (0, 1, 2, 3)
  
Is there any other way to determine `feature_info.reduction()` in advance?

Reply:
Can we slice features in `forward` instead of providing "out_indicies"? Otherwise, I would recommend using `pretrained=False` for the tmp model and maybe initialize it on the meta device to avoid double memory consumption.

Reply:
In `timm.create_model()`,  default is `pretrained=False`
I think initialize tmp model to `torch.device("meta")` is good

```python
self.model = timm.create_model(name, pretrained=False, features_only=True).to("meta")
```

what do you think?

Reply:
It should be something like this:

```python
with torch.device("meta"):
    self.model = timm.create_model(name, pretrained=False, features_only=True)
```

Reply:
Not all timm models support initialize on meta device, check on this PR:
https://github.com/huggingface/pytorch-image-models/pull/2350

Note: meta device need pytorch version >= 1.9

Reply:
We are safe for torch, the minimum right now is 1.9. Re meta device: do those model raise an error while initializing on meta device?

Reply:
Most transformer-style models raise error because `torch.linspace` not supported on meta device.

## test code
```python
import segmentation_models_pytorch as smp
model = smp.encoders.get_encoder('tu-convnextv2_tiny', weights=None)
```
### output
```
timm/models/convnext.py, line 331, in __init__
    dp_rates = [x.tolist() for x in torch.linspace(0, drop_path_rate, sum(depths)).split(depths)]
NotImplementedError: Cannot copy out of meta tensor; no data!
```

---

short test like this:

```python
import torch
depths = (3, 3, 9, 3)
drop_path_rate = 0.0
with torch.device("meta"):
    dp_rates = [x.tolist() for x in torch.linspace(0, drop_path_rate, sum(depths)).split(depths)]
```

Reply:
Ok, then let's leave it without meta  for now 👍

Reply:
Later it can be added with timm version constraint

Reply:
The issue occurs when initializing TimmUniversalEncoder, where it takes twice the time to build the model (tmp_model and self.model). If tmp_model can be created on the meta device, the time would be negligible, requiring only the time for self.model build.

Maybe we can use this method to accelerate create model for the time being.
```python
# Load a temporary model to analyze its feature hierarchy
try:
    with torch.device("meta"):
        self.model = timm.create_model(name, features_only=True)
except Exception:
    self.model = timm.create_model(name, features_only=True)
```

what do you think

Reply:
Another question is that we need to delete 'tmp_model' at the last?

Reply:
Sounds good 👍

Reply:
> Another question is that we need to delete tmp_model at the last?

it will be removed with garbage collector after init.. the only case we want to remove it explicitly - as soon as we got features info - to save memory, but I'm not sure if this is async or will slow down model initialization

Reply:
most probably

```python
del tmp_model
torch.cuda.empty_cache()
```

will release the memory, but will be slow

Reply:
lets leave it as is for now, it can be optimized later if needed

Reply:
As is I mean:

```
# Load a temporary model to analyze its feature hierarchy
try:
    with torch.device("meta"):
        tmp_model = timm.create_model(name, features_only=True)
except Exception:
    tmp_model = timm.create_model(name, features_only=True)
```

sorry for the confusuion

Reply:
> If we don't use additional variable names, it shouldn't take up extra memory?
don't think so, we still allocate twice.

1. we have tmp model initialized and linked to `self.model`
2. we initialize required model
3. we unlink tmp model from var name and link required one

two models exist at a time

Reply:
I think you are right.